From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sleigh-InSPECtor <sleighinspector@outlook.com>
Date: Thu, 23 May 2024 14:54:54 +0930
Subject: [PATCH] 6568: AArch32: (Thumb32) missing carry(CY) flag update with
 shift_carry

AArch32: (Thumb32) fixed shift_carry CY flag update

* numerous instructions had missing carry(CY) flag update with shift_carry
---
 .../data/languages/ARMTHUMBinstructions.sinc  | 41 +++++++++++--------
 1 file changed, 23 insertions(+), 18 deletions(-)

diff --git a/Ghidra/Processors/ARM/data/languages/ARMTHUMBinstructions.sinc b/Ghidra/Processors/ARM/data/languages/ARMTHUMBinstructions.sinc
index 0f0819bcc0..30a397f00d 100644
--- a/Ghidra/Processors/ARM/data/languages/ARMTHUMBinstructions.sinc
+++ b/Ghidra/Processors/ARM/data/languages/ARMTHUMBinstructions.sinc
@@ -264,6 +264,11 @@ macro resflags(result) {
   tmpZR = result == 0;
 }
 
+macro resflags_with_shift_carry(result) {
+  resflags(result);
+  tmpCY = shift_carry;
+}
+
 macro th_logicflags() {
   tmpCY = shift_carry;
   tmpOV = OV;
@@ -1260,22 +1265,22 @@ with : ARMcondCk=1 {
 
 @if defined(VERSION_6T2) || defined(VERSION_7)
 
-:and^thSBIT_ZN^ItCond	Rd0811,Rn0003,ThumbExpandImm12 		is TMode=1 & ItCond & (op11=0x1e & thc0909=0 & sop0508=0 & thSBIT_ZN & Rn0003; thc1515=0 & Rd0811) & ThumbExpandImm12
+:and^thSBIT_CZN^ItCond	Rd0811,Rn0003,ThumbExpandImm12 		is TMode=1 & ItCond & (op11=0x1e & thc0909=0 & sop0508=0 & thSBIT_CZN & Rn0003; thc1515=0 & Rd0811) & ThumbExpandImm12
 {
   build ItCond;
   build ThumbExpandImm12;
   Rd0811 = Rn0003 & ThumbExpandImm12;
-  resflags(Rd0811);
-  build thSBIT_ZN;
+  resflags_with_shift_carry(Rd0811);
+  build thSBIT_CZN;
 }
 
-:and^thSBIT_ZN^ItCond^".w"	Rd0811,Rn0003,thshift2 		is TMode=1 & ItCond & op11=0x1d & thc0910=1 & sop0508=0 & thSBIT_ZN & Rn0003; thc1515=0 & Rd0811 & thshift2
+:and^thSBIT_CZN^ItCond^".w"	Rd0811,Rn0003,thshift2 		is TMode=1 & ItCond & op11=0x1d & thc0910=1 & sop0508=0 & thSBIT_CZN & Rn0003; thc1515=0 & Rd0811 & thshift2
 {
   build ItCond;
   build thshift2;
   Rd0811 = Rn0003 & thshift2;
-  resflags(Rd0811);
-  build thSBIT_ZN;
+  resflags_with_shift_carry(Rd0811);
+  build thSBIT_CZN;
 }
 @endif # VERSION_6T2 || VERSION_7
 
@@ -1403,13 +1408,13 @@ macro th_set_carry_for_asr(op1,shift_count) {
 
 @if defined(VERSION_6T2) || defined(VERSION_7)
 
-:bic^thSBIT_ZN^ItCond   Rd0811,Rn0003,ThumbExpandImm12 		is TMode=1 & ItCond & (op11=0x1e & thc0909=0 & sop0508=1 & thSBIT_ZN & Rn0003; thc1515=0 & Rd0811) & ThumbExpandImm12
+:bic^thSBIT_CZN^ItCond   Rd0811,Rn0003,ThumbExpandImm12 		is TMode=1 & ItCond & (op11=0x1e & thc0909=0 & sop0508=1 & thSBIT_CZN & Rn0003; thc1515=0 & Rd0811) & ThumbExpandImm12
 {
   build ItCond;
   build ThumbExpandImm12;
   Rd0811 = Rn0003&(~ThumbExpandImm12);
-  resflags(Rd0811);
-  build thSBIT_ZN;
+  resflags_with_shift_carry(Rd0811);
+  build thSBIT_CZN;
 }
 
 :bic^thSBIT_CZNO^ItCond^".w"	Rd0811,Rn0003,thshift2 		is TMode=1 & ItCond & op11=0x1d & thc0910=1 & sop0508=1 & thSBIT_CZNO & Rn0003; thc1515=0 & Rd0811 & thshift2
@@ -2665,12 +2670,12 @@ macro th_set_carry_for_lsr(op1,shift_count) {
 
 @if defined(VERSION_6T2) || defined(VERSION_7)
 
-:mov^thSBIT_ZN^ItCond^".w"	Rd0811,ThumbExpandImm12 		is TMode=1 & ItCond & (op11=0x1e & thc0909=0 & sop0508=2 & thSBIT_ZN & sop0003=15; thc1515=0 & Rd0811) & ThumbExpandImm12
+:mov^thSBIT_CZN^ItCond^".w"	Rd0811,ThumbExpandImm12 		is TMode=1 & ItCond & (op11=0x1e & thc0909=0 & sop0508=2 & thSBIT_CZN & sop0003=15; thc1515=0 & Rd0811) & ThumbExpandImm12
 {
   build ItCond;
   Rd0811 = ThumbExpandImm12;
-  resflags(Rd0811);
-  build thSBIT_ZN;
+  resflags_with_shift_carry(Rd0811);
+  build thSBIT_CZN;
 }
 
 :movw^ItCond	Rd0811,Immed16 		is TMode=1 & ItCond & (op11=0x1e & thc0909=1 & sop0508=2 & thc0404=0; thc1515=0 & Rd0811) & Immed16
@@ -3012,20 +3017,20 @@ thspsrmask: spsr^thpsrmask	is thpsrmask & spsr { export thpsrmask; }
   spsr = (spsr& ~thspsrmask) | (Rn0003 & thspsrmask);
 }
 
-:mvn^thSBIT_ZN^ItCond    Rd0811,ThumbExpandImm12  is TMode=1 & ItCond & (op11=0x1e & thc0909=0 & sop0508=3 & thSBIT_ZN & thc0003=15; thc1515=0 & Rd0811) & ThumbExpandImm12
+:mvn^thSBIT_CZN^ItCond    Rd0811,ThumbExpandImm12  is TMode=1 & ItCond & (op11=0x1e & thc0909=0 & sop0508=3 & thSBIT_CZN & thc0003=15; thc1515=0 & Rd0811) & ThumbExpandImm12
 {
   build ItCond;
   Rd0811 = ~ThumbExpandImm12;
-  resflags(Rd0811);
-  build thSBIT_ZN;
+  resflags_with_shift_carry(Rd0811);
+  build thSBIT_CZN;
 }
 
-:mvn^thSBIT_ZN^ItCond^".w"  Rd0811,thshift2 		is TMode=1 & ItCond & op11=0x1d & thc0910=1 & sop0508=3 & thSBIT_ZN & thc0003=15; thc1515=0 & Rd0811 & thshift2
+:mvn^thSBIT_CZN^ItCond^".w"  Rd0811,thshift2 		is TMode=1 & ItCond & op11=0x1d & thc0910=1 & sop0508=3 & thSBIT_CZN & thc0003=15; thc1515=0 & Rd0811 & thshift2
 {
   build ItCond;
   Rd0811 = ~thshift2;
-  resflags(Rd0811);
-  build thSBIT_ZN;
+  resflags_with_shift_carry(Rd0811);
+  build thSBIT_CZN;
 }
 
 @endif # VERSION_6T2 || VERSION_7
-- 
2.45.1

